### PR TITLE
Feature/pre commit hook

### PR DESCRIPTION
**Pull request by frisitano**
I have run a first time pass of the following pre-commit configuration over the repository and addressed a number of static analysis issues.  

Pre-commit config:
`
repos:
  - repo: https://github.com/ambv/black
    rev: 20.8b1
    hooks:
      - id: black

  - repo: https://gitlab.com/pycqa/flake8
    rev: '3.8.4'
    hooks:
      - id: flake8

  - repo: https://github.com/pre-commit/mirrors-isort
    rev: v5.6.4
    hooks:
      - id: isort

  - repo: https://github.com/pycqa/pydocstyle
    rev: 5.1.1
    hooks:
    -   id: pydocstyle
        args: [--convention=google]

  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: 'v0.790'
    hooks:
    -   id: mypy
`

There are still a number of outstanding static analysis issues.  Primarily around doc strings but also an issue with create_app() complexity reported by flake8 and also a Name error from mypy.  See below:

`
pre-commit run --all-files
black....................................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

src/__init__.py:20:1: C901 'create_app' is too complex (22)

isort....................................................................Passed
pydocstyle...............................................................Failed
- hook id: pydocstyle
- exit code: 1

src/pql/parser.py:1 at module level:
        D100: Missing docstring in public module
src/pql/parser.py:12 in public class `Parser`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/parser.py:12 in public class `Parser`:
        D415: First line should end with a period, question mark, or exclamation point (not 'd')
src/pql/parser.py:68 in public function `parse_and_execute`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/parser.py:68 in public function `parse_and_execute`:
        D415: First line should end with a period, question mark, or exclamation point (not 'r')
tests/integration/test_psql.py:1 at module level:
        D100: Missing docstring in public module
tests/integration/test_psql.py:5 in public function `test_psql_connection`:
        D103: Missing docstring in public function
tests/integration/conftest.py:1 at module level:
        D100: Missing docstring in public module
tests/integration/conftest.py:7 in public function `app`:
        D103: Missing docstring in public function
tests/integration/conftest.py:13 in public function `client`:
        D103: Missing docstring in public function
src/logging.py:1 at module level:
        D100: Missing docstring in public module
src/logging.py:7 in public function `setup_logging`:
        D103: Missing docstring in public function
src/models/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/models/__init__.py:6 in public class `ContractOracle`:
        D101: Missing docstring in public class
src/network/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/network/accounts.py:1 at module level:
        D100: Missing docstring in public module
src/network/accounts.py:17 in public method `__init__`:
        D107: Missing docstring in __init__
src/network/accounts.py:21 in public method `load`:
        D205: 1 blank line required between summary line and description (found 0)
src/network/accounts.py:21 in public method `load`:
        D415: First line should end with a period, question mark, or exclamation point (not 'y')
examples/custom_function.py:1 at module level:
        D100: Missing docstring in public module
src/__main__.py:1 at module level:
        D100: Missing docstring in public module
src/__main__.py:26 in public function `main`:
        D103: Missing docstring in public function
src/pql/pipeline.py:1 at module level:
        D100: Missing docstring in public module
src/pql/pipeline.py:94 in public method `get_index`:
        D417: Missing argument descriptions in the docstring (argument(s) step_index are missing descriptions in 'get_index' docstring)
src/process/executor.py:1 at module level:
        D100: Missing docstring in public module
src/process/executor.py:20 in public function `handle_request_event`:
        D205: 1 blank line required between summary line and description (found 0)
src/process/executor.py:20 in public function `handle_request_event`:
        D415: First line should end with a period, question mark, or exclamation point (not 'd')
src/process/executor.py:20 in public function `handle_request_event`:
        D417: Missing argument descriptions in the docstring (argument(s) self are missing descriptions in 'handle_request_event' docstring)
src/migrations/versions/8bf91192e068_add_contractoracle_model.py:1 at module level:
        D415: First line should end with a period, question mark, or exclamation point (not 'l')
src/migrations/versions/8bf91192e068_add_contractoracle_model.py:18 in public function `upgrade`:
        D103: Missing docstring in public function
src/migrations/versions/8bf91192e068_add_contractoracle_model.py:29 in public function `downgrade`:
        D103: Missing docstring in public function
examples/ethereum_request.py:1 at module level:
        D100: Missing docstring in public module
examples/ethereum_request.py:8 in public function `main`:
        D205: 1 blank line required between summary line and description (found 0)
__init__.py:1 at module level:
        D104: Missing docstring in public package
tests/__init__.py:1 at module level:
        D104: Missing docstring in public package
docs/source/conf.py:1 at module level:
        D100: Missing docstring in public module
src/pql/exceptions.py:1 at module level:
        D100: Missing docstring in public module
src/pql/exceptions.py:8 in public class `PqlDecodingError`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/exceptions.py:8 in public class `PqlDecodingError`:
        D415: First line should end with a period, question mark, or exclamation point (not 's')
src/pql/exceptions.py:12 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:21 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:28 in public class `NoInputValue`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/exceptions.py:28 in public class `NoInputValue`:
        D415: First line should end with a period, question mark, or exclamation point (not 't')
src/pql/exceptions.py:32 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:41 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:50 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:59 in public method `__init__`:
        D107: Missing docstring in __init__
examples/simple_postgres_request.py:1 at module level:
        D100: Missing docstring in public module
src/utils/ipfs.py:1 at module level:
        D100: Missing docstring in public module
src/pql/handlers/handler.py:1 at module level:
        D100: Missing docstring in public module
src/pql/handlers/handler.py:11 in public method `execute`:
        D415: First line should end with a period, question mark, or exclamation point (not 'N')
src/pql/handlers/handler.py:23 in public method `require_params`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/handlers/eth_handler.py:1 at module level:
        D100: Missing docstring in public module
src/pql/handlers/eth_handler.py:20 in public method `execute`:
        D102: Missing docstring in public method
src/pql/handlers/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/cli/accounts.py:1 at module level:
        D100: Missing docstring in public module
src/cli/accounts.py:15 in public function `main`:
        D103: Missing docstring in public function
src/pql/handlers/sql_handler.py:1 at module level:
        D100: Missing docstring in public module
src/cli/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/cli/node.py:1 at module level:
        D100: Missing docstring in public module
src/cli/node.py:22 in public function `main`:
        D103: Missing docstring in public function
src/utils/__init__.py:1 at module level:
        D104: Missing docstring in public package
tests/integration/test_http.py:1 at module level:
        D100: Missing docstring in public module
tests/integration/test_http.py:6 in public function `test_http_get`:
        D103: Missing docstring in public function
tests/integration/test_http.py:47 in public function `test_http_post`:
        D103: Missing docstring in public function
examples/simple_get_request.py:1 at module level:
        D100: Missing docstring in public module
tests/integration/test_aggregate.py:1 at module level:
        D100: Missing docstring in public module
tests/integration/test_aggregate.py:9 in public function `pql_json`:
        D103: Missing docstring in public function
tests/integration/test_aggregate.py:64 in public function `mock_responses`:
        D103: Missing docstring in public function
tests/integration/test_aggregate.py:131 in public function `test_mean`:
        D103: Missing docstring in public function
examples/aggregate_bitcoin_price.py:1 at module level:
        D100: Missing docstring in public module
examples/ipfs_request.py:1 at module level:
        D100: Missing docstring in public module
src/pql/handlers/rest_api_handler.py:1 at module level:
        D100: Missing docstring in public module
src/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/__init__.py:20 in public function `create_app`:
        D103: Missing docstring in public function
src/migrations/env.py:1 at module level:
        D100: Missing docstring in public module
src/process/collector.py:1 at module level:
        D100: Missing docstring in public module
src/process/collector.py:18 in public function `listen_for_request_events`:
        D205: 1 blank line required between summary line and description (found 0)
src/process/collector.py:18 in public function `listen_for_request_events`:
        D415: First line should end with a period, question mark, or exclamation point (not 'd')
src/pql/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/network/web3.py:1 at module level:
        D100: Missing docstring in public module
src/config.py:1 at module level:
        D100: Missing docstring in public module
src/config.py:14 in public class `Config`:
        D101: Missing docstring in public class
src/config.py:51 in public method `__init__`:
        D107: Missing docstring in __init__
tests/integration/test_custom_function.py:1 at module level:
        D100: Missing docstring in public module
tests/integration/test_custom_function.py:10 in public function `pql`:
        D103: Missing docstring in public function
tests/integration/test_custom_function.py:38 in public function `test_custom_function_not_found`:
        D103: Missing docstring in public function
tests/integration/test_custom_function.py:64 in public function `test_custom_function`:
        D103: Missing docstring in public function
src/process/__init__.py:1 at module level:
        D104: Missing docstring in public package
tests/integration/__init__.py:1 at module level:
        D104: Missing docstring in public package

mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src/models/__init__.py:6: error: Name 'db.Model' is not defined
Found 1 error in 1 file (checked 42 source files)
`

I'm unsure how we would like to address the mypy error regarding "error:Name 'db.Model' is not defined".  I don't see a simple resolution.  Also would appreciate some guidance on the create_app complexity error reported by black.  We could force that both these errors are ignored by the static analysis tools, would this be a reasonable solution?  There are also a large number of docstring errors as seen above.  

How would we like to proceed.  We have 2 options:

1. Merge this PR and then any new PR's will have to be run through pre-commit to ensure new code is aligned with standards and then address the pre-existing static analysis on a rolling basis.
2. Fix all errors upfront, including docstrings, before we merge the PR.

### CODE REVIEW

File: src/migrations/env.py [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
 
 from alembic import context
 from sqlalchemy import engine_from_config, pool
 
-sys.path.append(".")
-
 from src.config import config as app_config
 from src.models import db
 
+sys.path.append(".")
```
Comment (jbargu) [inline src/migrations/env.py:11]: I believe this has to be before the `src.*` imports otherwise it can't import the packages.

**Comment by jbargu** [comment]
As per your suggestion I would also add an ignore statement for `create_app` for now. It is too early to optimize and the future PRs will probably refactor it many times.

I would ignore docstrings on `tests` folder. Writing tests is tedious as it is and usually the name of the test function is representative enough.

Regarding the `db.Model` it appears to be still [open issue](https://github.com/python/mypy/issues/8603) in `mypy`. Not sure if [ignoring](https://stackoverflow.com/questions/56774322/flask-sql-alchemy-vs-mypy-error-with-model-type) all non-fatal errors is the way to go. Could also ignore the line for the time being. However it will become a bigger problem later on with more models.

I would go for option 1 and we can weed out the warnings eventually. If you can please resolve the conflicts with the current master branch. In any case thank your for your great work!

**Comment by frisitano** [comment]
I've addressed the points raised and resolved the conflicts with master.  All hooks with the exception of pydocstyle are passing.  Here is the pre-commit log following these changes:

pre-commit run --all-files 
black....................................................................Passed
flake8...................................................................Passed
isort....................................................................Passed
pydocstyle...............................................................Failed
- hook id: pydocstyle
- exit code: 1

src/config.py:1 at module level:
        D100: Missing docstring in public module
src/config.py:14 in public class `Config`:
        D101: Missing docstring in public class
src/config.py:51 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/handlers/sql_handler.py:1 at module level:
        D100: Missing docstring in public module
src/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/__init__.py:20 in public function `create_app` (skipping C901):
        D103: Missing docstring in public function
src/network/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/models/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/models/__init__.py:6 in public class `ContractOracle`:
        D101: Missing docstring in public class
src/logging.py:1 at module level:
        D100: Missing docstring in public module
src/logging.py:7 in public function `setup_logging`:
        D103: Missing docstring in public function
src/migrations/env.py:1 at module level:
        D100: Missing docstring in public module
__init__.py:1 at module level:
        D104: Missing docstring in public package
src/pql/handlers/eth_handler.py:1 at module level:
        D100: Missing docstring in public module
src/pql/handlers/eth_handler.py:20 in public method `execute`:
        D102: Missing docstring in public method
examples/custom_function.py:1 at module level:
        D100: Missing docstring in public module
src/pql/parser.py:1 at module level:
        D100: Missing docstring in public module
src/pql/parser.py:13 in public class `Parser`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/parser.py:13 in public class `Parser`:
        D415: First line should end with a period, question mark, or exclamation point (not 'd')
src/pql/parser.py:78 in public function `parse_and_execute`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/parser.py:78 in public function `parse_and_execute`:
        D415: First line should end with a period, question mark, or exclamation point (not 'r')
src/utils/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/process/__init__.py:1 at module level:
        D104: Missing docstring in public package
examples/ipfs_request.py:1 at module level:
        D100: Missing docstring in public module
docs/source/conf.py:1 at module level:
        D100: Missing docstring in public module
src/cli/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/pql/handlers/rest_api_handler.py:1 at module level:
        D100: Missing docstring in public module
src/__main__.py:1 at module level:
        D100: Missing docstring in public module
src/__main__.py:26 in public function `main`:
        D103: Missing docstring in public function
src/cli/accounts.py:1 at module level:
        D100: Missing docstring in public module
src/cli/accounts.py:15 in public function `main`:
        D103: Missing docstring in public function
examples/user_defined_sql_aggregation.py:1 at module level:
        D100: Missing docstring in public module
src/pql/exceptions.py:1 at module level:
        D100: Missing docstring in public module
src/pql/exceptions.py:8 in public class `PqlDecodingError`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/exceptions.py:8 in public class `PqlDecodingError`:
        D415: First line should end with a period, question mark, or exclamation point (not 's')
src/pql/exceptions.py:12 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:21 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:28 in public class `NoInputValue`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/exceptions.py:28 in public class `NoInputValue`:
        D415: First line should end with a period, question mark, or exclamation point (not 't')
src/pql/exceptions.py:32 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:41 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:50 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:59 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:68 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:77 in public method `__init__`:
        D107: Missing docstring in __init__
src/pql/exceptions.py:86 in public method `__init__`:
        D107: Missing docstring in __init__
examples/simple_get_request.py:1 at module level:
        D100: Missing docstring in public module
src/pql/handlers/handler.py:1 at module level:
        D100: Missing docstring in public module
src/pql/handlers/handler.py:11 in public method `execute`:
        D415: First line should end with a period, question mark, or exclamation point (not 'N')
src/pql/handlers/handler.py:23 in public method `require_params`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/handlers/__init__.py:1 at module level:
        D104: Missing docstring in public package
examples/aggregate_bitcoin_price.py:1 at module level:
        D100: Missing docstring in public module
src/pql/__init__.py:1 at module level:
        D104: Missing docstring in public package
src/process/executor.py:1 at module level:
        D100: Missing docstring in public module
src/process/executor.py:20 in public function `handle_request_event`:
        D205: 1 blank line required between summary line and description (found 0)
src/process/executor.py:20 in public function `handle_request_event`:
        D415: First line should end with a period, question mark, or exclamation point (not 'd')
src/process/executor.py:20 in public function `handle_request_event`:
        D417: Missing argument descriptions in the docstring (argument(s) self are missing descriptions in 'handle_request_event' docstring)
src/process/collector.py:1 at module level:
        D100: Missing docstring in public module
src/process/collector.py:18 in public function `listen_for_request_events`:
        D205: 1 blank line required between summary line and description (found 0)
src/process/collector.py:18 in public function `listen_for_request_events`:
        D415: First line should end with a period, question mark, or exclamation point (not 'd')
examples/ethereum_request.py:1 at module level:
        D100: Missing docstring in public module
examples/ethereum_request.py:8 in public function `main`:
        D205: 1 blank line required between summary line and description (found 0)
src/migrations/versions/8bf91192e068_add_contractoracle_model.py:1 at module level:
        D415: First line should end with a period, question mark, or exclamation point (not 'l')
src/migrations/versions/8bf91192e068_add_contractoracle_model.py:18 in public function `upgrade`:
        D103: Missing docstring in public function
src/migrations/versions/8bf91192e068_add_contractoracle_model.py:29 in public function `downgrade`:
        D103: Missing docstring in public function
examples/simple_postgres_request.py:1 at module level:
        D100: Missing docstring in public module
src/pql/query_sql.py:1 at module level:
        D100: Missing docstring in public module
src/pql/query_sql.py:57 in public function `construct_aggregate_sql_payload`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/query_sql.py:57 in public function `construct_aggregate_sql_payload`:
        D415: First line should end with a period, question mark, or exclamation point (not 'r')
src/pql/query_sql.py:57 in public function `construct_aggregate_sql_payload`:
        D403: First word of the first line should be properly capitalized ('Parses', not 'parses')
src/network/web3.py:1 at module level:
        D100: Missing docstring in public module
src/network/accounts.py:1 at module level:
        D100: Missing docstring in public module
src/network/accounts.py:17 in public method `__init__`:
        D107: Missing docstring in __init__
src/network/accounts.py:21 in public method `load`:
        D205: 1 blank line required between summary line and description (found 0)
src/network/accounts.py:21 in public method `load`:
        D415: First line should end with a period, question mark, or exclamation point (not 'y')
src/cli/node.py:1 at module level:
        D100: Missing docstring in public module
src/cli/node.py:22 in public function `main`:
        D103: Missing docstring in public function
src/pql/pipeline.py:1 at module level:
        D100: Missing docstring in public module
src/pql/pipeline.py:97 in public method `get_index`:
        D417: Missing argument descriptions in the docstring (argument(s) step_index are missing descriptions in 'get_index' docstring)
src/pql/pipeline.py:141 in public method `query_sql`:
        D205: 1 blank line required between summary line and description (found 0)
src/pql/pipeline.py:141 in public method `query_sql`:
        D210: No whitespaces allowed surrounding docstring text
src/pql/pipeline.py:141 in public method `query_sql`:
        D415: First line should end with a period, question mark, or exclamation point (not 'e')
src/utils/ipfs.py:1 at module level:
        D100: Missing docstring in public module

mypy.....................................................................Passed

**Comment by frisitano** [comment]
Having reviewed the log I was thinking we may be better off removing the pydocstyle hook from the automated pre-commit hook and running it manually for the time being until we can fix all the errors?  This will allow for all green pre-commit runs.  What do you think @jbargu?

**Comment by jbargu** [comment]
Yeah good point, I'll merge this PR and remove the hook! Awesome!

**Pull request merged**